src/cursor/cur_std.c

```diff
@@ -556,6 +581,215 @@ err:		cursor->saved_err = ret;
+		    cursor->uri, strlen(cursor->uri)) % WT_HASH_ARRAY_SIZE;
+	TAILQ_REMOVE(&session->cursors, cursor, q);
+	TAILQ_INSERT_HEAD(
+	    &session->cursor_cache[cursor->cache_bucket], cursor, q);
+
+	(void)__wt_atomic_sub32(&S2C(session)->open_cursor_count, 1);
+	WT_STAT_DATA_DECR(session, session_cursor_open);
+	WT_STAT_DATA_INCR(session, session_cursor_cached);
+	F_SET(cursor, WT_CURSTD_CACHED);
+	return (ret);
+}
+
+/*
+ * __wt_cursor_reopen --
+ *	Reopen this cursor from the cached state.
+ */
+void
+__wt_cursor_reopen(WT_CURSOR *cursor, WT_DATA_HANDLE *dhandle)
+{
+	WT_SESSION_IMPL *session;
+
+	session = (WT_SESSION_IMPL *)cursor->session;
+	WT_ASSERT(session, F_ISSET(cursor, WT_CURSTD_CACHED));
+
+	if (dhandle != NULL) {
+		session->dhandle = dhandle;
+		__wt_cursor_dhandle_incr_use(session);
+		WT_DHANDLE_RELEASE(dhandle);
+	}
+	(void)__wt_atomic_add32(&S2C(session)->open_cursor_count, 1);
+	WT_STAT_DATA_INCR(session, session_cursor_open);
+	WT_STAT_DATA_DECR(session, session_cursor_cached);
+
+	TAILQ_REMOVE(&session->cursor_cache[cursor->cache_bucket], cursor, q);
+	TAILQ_INSERT_HEAD(&session->cursors, cursor, q);
+	F_CLR(cursor, WT_CURSTD_CACHED);
+}
+
+/*
+ * __wt_cursor_cache_release --
+ *	Put the cursor into a cached state, called during close operations.
+ */
+int
+__wt_cursor_cache_release(WT_SESSION_IMPL *session, WT_CURSOR *cursor,
+    bool *released)
+{
+	WT_DECL_RET;
+
+	*released = false;
+	if (!F_ISSET(cursor, WT_CURSTD_CACHEABLE) ||
+	    !F_ISSET(session, WT_SESSION_CACHE_CURSORS))
+		return (0);
+
+	WT_ASSERT(session, !F_ISSET(cursor, WT_CURSTD_BULK | WT_CURSTD_CACHED));
+
+	/*
+	 * Do any sweeping first, if there are errors, it will
+	 * be easier to clean up if the cursor is not already cached.
+	 */
+	if (--session->cursor_sweep_countdown == 0)
+		WT_RET(__wt_session_cursor_cache_sweep(session));
+
+	WT_ERR(cursor->cache(cursor));
+	WT_STAT_CONN_INCR(session, cursor_cache);
+	WT_STAT_DATA_INCR(session, cursor_cache);
+	WT_ASSERT(session, F_ISSET(cursor, WT_CURSTD_CACHED));
+	*released = true;
+
+	if (0) {
+		/*
+		 * If caching fails, we must restore the state of the
+		 * cursor back to open so that the close works from
+		 * a known state. The reopen may also fail, but that
+		 * doesn't matter at this point.
+		 */
+err:		WT_TRET(cursor->reopen(cursor, false));
+		WT_ASSERT(session, !F_ISSET(cursor, WT_CURSTD_CACHED));
+	}
+
+	return (ret);
+}
+
+/*
+ * __wt_cursor_cache_get --
+ *	Open a matching cursor from the cache.
+ */
+int
+__wt_cursor_cache_get(WT_SESSION_IMPL *session, const char *uri,
+    WT_CURSOR *owner, const char *cfg[], WT_CURSOR **cursorp)
+{
+	WT_CONFIG_ITEM cval;
+	WT_CONFIG_ITEM_STATIC_INIT(false_value);
+	WT_CURSOR *cursor;
+	WT_DECL_RET;
+	uint64_t bucket;
+	bool have_config;
+
+	if (owner != NULL && F_ISSET(owner, WT_CURSTD_CACHEABLE))
+		return (WT_NOTFOUND);
+	have_config = (cfg != NULL && cfg[0] != NULL && cfg[1] != NULL);
+	if (have_config) {
+		WT_RET(__wt_config_gets_def(
```

Comment:
All of these that do the same thing should be alphabetized on the config string name I think. Also a small comment about not ever using the cache for specially configured cursors would be helpful.

Reply:
Yes, I've clarified with some comments.